From: SeongJae Park <sj@kernel.org>
Date: Wed, 3 Sep 2025 18:08:43 -0700
Subject: [PATCH] mm/damon/core: reduce total_charged_{sz,ns} to avoid
 overflows

If time quota is being used and
total_charged_sz or total_charged_ns overflows, the scheme's
throughput calculation can

Signed-off-by: SeongJae Park <sj@kernel.org>
---
 mm/damon/core.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/mm/damon/core.c b/mm/damon/core.c
index 46d5479859fd..b677c975ea12 100644
--- a/mm/damon/core.c
+++ b/mm/damon/core.c
@@ -2305,6 +2305,10 @@ static void damos_adjust_quota(struct damon_ctx *c, struct damos *s)
 		if (quota->esz && quota->charged_sz >= quota->esz)
 			s->stat.qt_exceeds++;
 		quota->total_charged_sz += quota->charged_sz;
+		if (quota->total_charged_sz > ULONG_MAX / 2) {
+			quota->total_charged_sz /= 2;
+			quota->total_charged_ns /= 2;
+		}
 		quota->charged_from = jiffies;
 		quota->charged_sz = 0;
 		if (trace_damos_esz_enabled())
-- 
2.39.5

